nexus/reconfigurator/execution/src/clickhouse.rs

```diff
@@ -0,0 +1,475 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Deployment of Clickhouse keeper and server nodes via clickhouse-admin running in
+//! deployed clickhouse zones.
+
+use anyhow::anyhow;
+use camino::Utf8PathBuf;
+use clickhouse_admin_api::KeeperConfigurableSettings;
+use clickhouse_admin_api::ServerConfigurableSettings;
+use clickhouse_admin_client::Client;
+use clickhouse_admin_types::config::ClickhouseHost;
+use clickhouse_admin_types::config::RaftServerSettings;
+use clickhouse_admin_types::KeeperSettings;
+use clickhouse_admin_types::ServerSettings;
+use futures::future::Either;
+use futures::stream::FuturesUnordered;
+use futures::stream::StreamExt;
+use nexus_db_queries::context::OpContext;
+use nexus_types::deployment::BlueprintZonesConfig;
+use nexus_types::deployment::ClickhouseClusterConfig;
+use omicron_common::address::CLICKHOUSE_ADMIN_PORT;
+use omicron_uuid_kinds::OmicronZoneUuid;
+use omicron_uuid_kinds::SledUuid;
+use slog::error;
+use slog::info;
+use slog::warn;
+use std::collections::BTreeMap;
+use std::net::Ipv6Addr;
+use std::net::SocketAddr;
+use std::net::SocketAddrV6;
+use std::str::FromStr;
+
+const CLICKHOUSE_SERVER_CONFIG_DIR: &str =
+    "/opt/oxide/clickhouse_server/config.d";
+const CLICKHOUSE_KEEPER_CONFIG_DIR: &str =
+    "/opt/oxide/clickhouse_keeper/config.d";
+const CLICKHOUSE_DATA_DIR: &str = "/data";
+
+pub(crate) async fn deploy_nodes(
+    opctx: &OpContext,
+    zones: &BTreeMap<SledUuid, BlueprintZonesConfig>,
+    clickhouse_cluster_config: &ClickhouseClusterConfig,
+) -> Result<(), Vec<anyhow::Error>> {
+    let keeper_configs = match keeper_configs(zones, clickhouse_cluster_config)
+    {
+        Ok(keeper_configs) => keeper_configs,
+        Err(e) => {
+            // We can't proceed if we fail to generate configs.
+            // Let's be noisy about it.
+            error!(
+                opctx.log,
+                "failed to generate clickhouse keeper configs: {e}"
+            );
+            return Err(vec![e]);
+        }
+    };
+
+    let keeper_hosts: Vec<_> = keeper_configs
+        .iter()
+        .map(|s| ClickhouseHost::Ipv6(s.settings.listen_addr))
+        .collect();
+
+    let server_configs =
+        match server_configs(zones, clickhouse_cluster_config, keeper_hosts) {
+            Ok(server_configs) => server_configs,
+            Err(e) => {
+                // We can't proceed if we fail to generate configs.
+                // Let's be noisy about it.
+                error!(
+                    opctx.log,
+                    "Failed to generate clickhouse server configs: {e}"
+                );
+                return Err(vec![e]);
+            }
+        };
+
+    let mut errors = vec![];
+    let log = opctx.log.clone();
+
+    // Inform each clickhouse-admin server in a keeper zone or server zone about
+    // its node's configuration
+    let mut futs = FuturesUnordered::new();
```

Comment:
Only thing I'd add here is to consider restricting concurrency via `buffer_unordered`.